axiom: add kernel I2C firmware flashing module for Axiom AX80A

varunee22b069 opened the pull request:
This PR adds firmware and configuration flashing support for the
Axiom AX80A touch device via I2C in Linux.

- Implements a character device driver with ioctl support
- Uses request_firmware() for loading binary files
- Adds AXIOM_FW_UPDATE, CFG_UPDATE, and checksum retrieval
- Includes test/axiom_test.c: user-space utility to validate flashing
- README updated with build, usage, and test instructions

Tested on AX80A, with firmware/configuration applied successfully.

This is a vendor-specific implementation. Future work will
focus on generalizing the module for multiple Axiom models
and submitting to upstream kernel if applicable.

Commit:
axiom: add kernel I2C firmware flashing module for Axiom AX310

Adds a Linux kernel module to flash firmware and configuration
to the Axiom AX310 touch controller over I2C, using a character
device and ioctl interface.

- Creates /dev/axiom_device
- Supports:
  - AXIOM_FW_UPDATE / AXIOM_CFG_UPDATE
  - AXIOM_FW_CHECKSUM / AXIOM_CFG_CHECKSUM
- Uses request_firmware() for loading .alc/.bin files
- Handles chunked download and checksum verification
- Ported from vendor's Python-based flashing tool

Also includes:
- axiom_test.c in test/: simple ioctl-based user-space utility
- Updated README with test instructions and firmware layout

marksat commented:
Hi @varunee22b069,

Just having a quick look at it now, the first problem I have encountered is that when I compile it, it fails with the following message:

```
tnx@rpi4:~/aXiom_Linux_Kernel_Module/axiom_i2c_fw_flash $ make                                                                                                 
make -C /lib/modules/6.6.62+rpt-rpi-v8/build/ M=/home/tnx/aXiom_Linux_Kernel_Module/axiom_i2c_fw_flash modules                                                 
make[1]: Entering directory '/usr/src/linux-headers-6.6.62+rpt-rpi-v8'                                                                                         
  CC [M]  /home/tnx/aXiom_Linux_Kernel_Module/axiom_i2c_fw_flash/axiom_core.o                                                                                  
  CC [M]  /home/tnx/aXiom_Linux_Kernel_Module/axiom_i2c_fw_flash/axiom_i2c.o                                                                                   
/home/tnx/aXiom_Linux_Kernel_Module/axiom_i2c_fw_flash/axiom_i2c.c:296:18: error: initialization of ‘int (*)(struct i2c_client *)’ from incompatible pointer type ‘int (*)(struct i2c_client *, const struct i2c_device_id *)’ [-Werror=incompatible-pointer-types]                                                           
  296 |         .probe = axiom_i2c_probe,                                                                                                                      
      |                  ^~~~~~~~~~~~~~~                                                                                                                       
/home/tnx/aXiom_Linux_Kernel_Module/axiom_i2c_fw_flash/axiom_i2c.c:296:18: note: (near initialization for ‘axiom_i2c_driver.probe’)                            
/home/tnx/aXiom_Linux_Kernel_Module/axiom_i2c_fw_flash/axiom_i2c.c:297:19: error: initialization of ‘void (*)(struct i2c_client *)’ from incompatible pointer type ‘int (*)(struct i2c_client *)’ [-Werror=incompatible-pointer-types]                                                                                        
  297 |         .remove = axiom_i2c_remove,                                                                                                                    
      |                   ^~~~~~~~~~~~~~~~                                                                                                                     
/home/tnx/aXiom_Linux_Kernel_Module/axiom_i2c_fw_flash/axiom_i2c.c:297:19: note: (near initialization for ‘axiom_i2c_driver.remove’)                           
cc1: some warnings being treated as errors                                                                                                                     
make[3]: *** [/usr/src/linux-headers-6.6.62+rpt-common-rpi/scripts/Makefile.build:248: /home/tnx/aXiom_Linux_Kernel_Module/axiom_i2c_fw_flash/axiom_i2c.o] Error 1                                                                                                                                                            
make[2]: *** [/usr/src/linux-headers-6.6.62+rpt-common-rpi/Makefile:1946: /home/tnx/aXiom_Linux_Kernel_Module/axiom_i2c_fw_flash] Error 2                      
make[1]: *** [/usr/src/linux-headers-6.6.62+rpt-common-rpi/Makefile:246: __sub-make] Error 2                                                                   
make[1]: Leaving directory '/usr/src/linux-headers-6.6.62+rpt-rpi-v8'                                                                                          
make: *** [Makefile:8: default] Error 2
```

When I build the normal aXiom kernel module driver, it doesn't have an issue. So I suspect the axiom_i2c_probe() and axiom_i2c_remove() calls need to be updated to match.

marksat commented on axiom_i2c_fw_flash/axiom_i2c.c:
```diff
@@ -0,0 +1,304 @@
+
+// purpose: Function called in IRQ context when device is plugged in.
+// returns: Error code
+static int axiom_i2c_probe(struct i2c_client *i2cClient, const struct i2c_device_id *id)
```
As per the comment in the PR, the axiom_i2c_probe() function and axiom_i2c_remove() functions needs to be updated. I think maybe you forked from and older version of the repo.

These changes are already applied to the main code. See changes from this PR:
https://github.com/TouchNetix/aXiom_Linux_Kernel_Module/pull/11/files

varunee22b069 replied:
Updated the code to include kernel version checks and pushed the fix. Will follow up shortly with a cleanup commit to align with existing code structure.
Thank you for pointing it out.

varunee22b069 committed:
fix: restore kernel version check for i2c probe/remove compatibility

Missed version check while importing vendor code.
This adds kernel version checks to fix build failure on newer kernels.

Commit:
Refactor firmware flashing support for improved integration

- Removed separate axiom_i2c_fw directory and merged all relevant code into the root directory to avoid redundancy and simplify diffs.
- Eliminated separate Makefile; integrated firmware build into the root-level Makefile. Module can now be built directly using the existing build system.
- Added kernel version checks wherever required for compatibility.
- Removed unrelated or incomplete additions unrelated to firmware flashing.
- Ensured minimal changes to existing files to improve maintainability and upstream compatibility.

varunee22b069 commented:
Updated the structure to better align with the existing codebase and reduce redundancy. All version-specific handling is now in place, and unnecessary additions have been removed to keep things minimal and maintainable. The module can now be built directly using the root Makefile.

Note: The test .bin file referenced in the README wasn’t committed due to .gitignore. It was just for convenience, let me know if you'd prefer it included.
